### PR TITLE
Re-introduce removed lock for building buildSrc

### PR DESCRIPTION
**Pull request by wolfs**
This allows running two builds in parallel on the same project.

As an improvement, the lock is only released when the artifacts produced by build src are stored in a cache to be re-used by the root build.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSourceBuilder.java [hunk 1]
```diff
@@ -93,14 +105,29 @@ public String getBuildPath() {
...
+            }
         } finally {
-            buildController.stop();
+            buildSrcCache.close();
         }
     }
 
+    PersistentCache createCache(StartParameter startParameter) {
+        return cacheRepository
+            .cache(new File(startParameter.getCurrentDir(), ".gradle/noVersion/buildSrc"))
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSourceBuilder.java]: ⭕️ Why is there the word "noVersion" here when further down we put the gradle.version into a properties file?

Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSourceBuilder.java]: The location of the cache is always the same. The cache only gets invalidated when the properties are different. Locking still works.
Some time ago we used that information to decide whether to run `clean` on buildSrc or not (when the Gradle version changed).

**Comment by wolfs** [comment]
Yeah, the "Hey, we don't need the cache here" is exactly what happened to me when I removed it.

File: subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSourceBuilder.java [hunk 2]
```diff
@@ -93,14 +105,29 @@ public String getBuildPath() {
     }
 
     private ClassPath buildBuildSrc(StartParameter startParameter) {
-        BuildController buildController = createBuildController(startParameter);
+        final PersistentCache buildSrcCache = createCache(startParameter);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSourceBuilder.java]: If we only use this as a lock, should we reflect that in the types being used? What we are creating here is a closable lock, not a cache. I wonder how much refactor that would require.

**Comment by wolfs** [comment]
I am using a simple lock now, not a cache anymore.
I also adjusted the test, so it would always fail if we do not use a lock for `buildSrc`.

**Pull request merged**